### PR TITLE
[SE-0198] Introduced CustomPlaygroundDisplayConvertible and deprecated PlaygroundQuickLook/CustomPlaygroundQuickLookable

### PR DESCRIPTION
**Pull request by cwakamo**
Introduced the `CustomPlaygroundDisplayConvertible` protocol and unconditionally deprecated `PlaygroundQuickLook` and `CustomPlaygroundQuickLookable` (and all conformances to those protocols) as part of implementation of SE-0198.

These are the Swift-side changes for &lt;rdar://problem/32905293&gt;.

### CODE REVIEW

**Comment by cwakamo** [comment]
Getting this up now so I can do this:

@swift-ci Please test

**Comment by cwakamo** [comment]
@airspeedswift Would you mind reviewing this to make sure I'm not off-base with how I put this in and the wording of the deprecation warning? (Or, alternatively, suggest someone else who can review it instead?)

File: stdlib/public/SDK/AppKit/AppKit.swift [hunk 1]
```diff
@@ -14,6 +14,7 @@ import Foundation
 @_exported import AppKit
 
 extension NSCursor : _DefaultCustomPlaygroundQuickLookable {
+  @available(*, deprecated, message: "NSCursor._defaultCustomPlaygroundQuickLook will be removed in a future Swift version")
```
Comment (moiseev) [inline stdlib/public/SDK/AppKit/AppKit.swift]: I think we should still allow using it in Swift 3 and Swift 4.1 without warning. Therefore the attribute should look like:
`@available(swift, deprecated: 4.1, message: "...")`.

File: stdlib/public/core/Mirror.swift [hunk 2]
```diff
@@ -660,8 +660,22 @@ internal extension Mirror {
 
 //===--- QuickLooks -------------------------------------------------------===//
 
-/// The sum of types that can be used as a Quick Look representation.
```
Comment (natecook1000) [inline stdlib/public/core/Mirror.swift]: Let's keep the old abstract here and in the next one—the `@availability` deprecation gives enough information to mark the symbol.

File: stdlib/public/core/Mirror.swift [hunk 3]
```diff
@@ -660,8 +660,22 @@ internal extension Mirror {
...
+/// `PlaygroundQuickLook`, or conditionalize your use such that it is only
+/// present when compiling with Swift 4.0 or Swift 3.2 or earlier:
+///
+/// ```swift
+/// #if !(swift(>=4.1) || swift(>=3.3) && !swift(>=4.0))
+///   /* OK to use PlaygroundQuickLook */
+/// #endif
+/// ```
+///
+/// - seealso: `CustomPlaygroundDisplayConvertible`
```
Comment (natecook1000) [inline stdlib/public/core/Mirror.swift]: Can you drop these `seealso` markers, too?

File: stdlib/public/core/Mirror.swift [hunk 4]
```diff
@@ -660,8 +660,22 @@ internal extension Mirror {
...
-/// The sum of types that can be used as a Quick Look representation.
+/// `PlaygroundQuickLook` is deprecated, and will be removed from the standard
+/// library in a future Swift release. Customizing display for in a playground
+/// is now done using the `CustomPlaygroundDisplayConvertible` protocol, which
+/// does not use the `PlaygroundQuickLook` enum. Please remove your uses of
+/// `PlaygroundQuickLook`, or conditionalize your use such that it is only
+/// present when compiling with Swift 4.0 or Swift 3.2 or earlier:
+///
+/// ```swift
+/// #if !(swift(>=4.1) || swift(>=3.3) && !swift(>=4.0))
```
Comment (natecook1000) [inline stdlib/public/core/Mirror.swift]: Our Markdown style for examples is to use indented code blocks, not fenced ones, unless we're writing non-Swift code. The code itself should be indented with 4 spaces.

File: stdlib/public/core/PlaygroundDisplay.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+///   `CustomPlaygroundDisplayConvertible`, then playground logging will ask for
+///   that instance's `playgroundDescription` and so on. It is permissible for
+///   playground logging implementations to place a reasonable limit on this
+///   kind of chaining to prevent infinite loops.
+public protocol CustomPlaygroundDisplayConvertible {
+    /// Returns the custom playground description for this instance.
+    ///
+    /// If this type has value semantics, the instance returned should be
+    /// unaffected by subsequent mutations if possible.
+    var playgroundDescription: Any { get }
```
Comment (natecook1000) [inline stdlib/public/core/PlaygroundDisplay.swift]: This block should only be indented two spaces.

**Comment by cwakamo** [comment]
@swift-ci Please smoke test

**Comment by cwakamo** [comment]
@natecook1000 I added back the old doc comments but left the deprecation notices as notes so that I could show how to properly conditionalize usage of `CustomPlaygroundQuickLookable`/`PlaygroundQuickLook` so that code continues to compile when we're able to remove them from the standard library. If you think that's overkill, I can drop my additions to those comments entirely.

**Review by natecook1000 — APPROVED** [review]
@cwakamo LGTM, thanks!

**Review by airspeedswift — APPROVED** [review]

**Pull request merged**